lib/di/provider/dio_provider.dart

```diff
@@ -9,14 +9,20 @@ const String defaultContentType = 'application/json; charset=utf-8';
 
 @LazySingleton()
 class DioProvider {
-  Dio? _dio;
+  Dio? _unauthenticatedDio;
+  Dio? _authenticatedDio;
   final LocalStorage _localStorage;
 
   DioProvider(this._localStorage);
 
-  Dio getDio() {
-    _dio ??= _createDio();
-    return _dio!;
+  Dio getUnAuthenticatedDio() {
+    _unauthenticatedDio ??= _createDio();
+    return _unauthenticatedDio!;
+  }
+
+  Dio getAuthenticatedDio() {
+    _authenticatedDio ??= _createDio(requireAuthenticate: true);
+    return _authenticatedDio!;
```

Comment:
What about pass parameter as default is false like we did in `_createDio`?
```suggestion
  Dio getDio({bool requireAuthenticate = false}) {
    _dio ??= _createDio(requireAuthenticate: requireAuthenticate);
    return _dio!;
  }
```

Reply:
@Wadeewee we need two separate `Dio` instances. For example, if `_dio` is already initialized by an unauthenticated instance, then we cannot create an authenticated instance with this approach because `_dio` will be non-null when calling `_dio ??= _createDio(requireAuthenticate: requireAuthenticate);`.

Reply:
> then we cannot create an authenticated instance with this approach because _dio will be non-null when calling _dio ??= _createDio(requireAuthenticate: requireAuthenticate);

@AVI5HEK May I know why we can't use this `approach` Could you please show me some error 🙏

Reply:
@Wadeewee `dio_provider` is a singleton. Consider these scenarios:
1. User is logged out. The **unauthenticated** instance of `Dio` will be used when calling the Login API.
2. When the login is successful, for all the subsequent calls, we'll need the **authenticated** instance. But since `_dio` will be non-null, calling `getDio()` will give us the **unauthenticated** instance of `Dio` — the method `_createDio(requireAuthenticate: requireAuthenticate)` will never be called.

Reply:
Your right @AVI5HEK 👍  I just finished debugging. It will initialize only once a time when building the app so we need to separate them 🚀